File: api/supabase_url.py
"""Helpers for normalizing Supabase project URLs."""
from __future__ import annotations


def normalize_supabase_url(url: str) -> str:
    value = url.strip().rstrip("/")
    suffixes = ("/auth/v1", "/rest/v1", "/storage/v1", "/functions/v1")

    changed = True
    while changed:
        changed = False
        lowered = value.lower()
        for suffix in suffixes:
            if lowered.endswith(suffix):
                value = value[: -len(suffix)].rstrip("/")
                changed = True
                break

    return value

def _shape_session_viewport_row(row: dict[str, object]) -> dict[str, object]:
    shaped = dict(row)
    payload = shaped.get('payload') or shaped.get('data') or {}
    if isinstance(payload, dict):
        shaped['payload'] = {key: value for key, value in payload.items() if value not in (None, '')}
    name = shaped.get('name') or shaped.get('title')
    if isinstance(name, str):
        shaped['name'] = name.strip()
    return shaped


def _shape_session_viewport_rows(rows: list[dict[str, object]]) -> list[dict[str, object]]:
    return [_shape_session_viewport_row(row) for row in rows]

File: web/lib/supabase/supabase-url.ts
/** Supabase clients expect the project root (https://ref.supabase.co), not .../auth/v1 or .../rest/v1. */
export function normalizeSupabaseUrl(url: string): string {
  let u = url.trim().replace(/\/+$/, "");
  const suffixes = ["/auth/v1", "/rest/v1", "/storage/v1", "/functions/v1"];
  let changed = true;
  while (changed) {
    changed = false;
    for (const s of suffixes) {
      if (u.toLowerCase().endsWith(s.toLowerCase())) {
        u = u.slice(0, -s.length).replace(/\/+$/, "");
        changed = true;
      }
    }
  }
  return u;
}
